lib/config_cat/config/user_comparator.ex

```diff
@@ -0,0 +1,525 @@
+    with {:ok, user_version} <- to_version(user_value),
+         {:ok, comparison_version} <- to_version(comparison_value) do
+      result = Version.compare(user_version, comparison_version)
+      {:ok, result in valid_comparisons}
+    end
+  end
+
+  defp compare_numbers(user_value, comparison_value, operator) do
+    with {:ok, user_float} <- to_float(user_value),
+         {:ok, comparison_float} <- to_float(comparison_value) do
+      {:ok, operator.(user_float, comparison_float)}
+    end
+  end
+
+  defp compare_datetimes(user_value, comparison_value, valid_comparisons) do
+    with {:ok, user_seconds} <- to_unix_seconds(user_value),
+         {:ok, comparison_seconds} <- to_float(comparison_value) do
+      result =
+        cond do
+          user_seconds < comparison_seconds -> :lt
+          user_seconds > comparison_seconds -> :gt
+          true -> :eq
+        end
+
+      {:ok, result in valid_comparisons}
+    else
+      {:error, :invalid_float} -> {:error, :invalid_datetime}
+    end
+  end
+
+  defp hash_value(value, %ComparisonContext{} = context) do
+    salted = value <> context.salt <> context.context_salt
+
+    :sha256
+    |> :crypto.hash(salted)
+    |> Base.encode16()
+    |> String.downcase()
+  end
+
+  defp parse_comparison(value) do
+    [length_string, comparison_string] = String.split(value, "_", parts: 2)
+
+    {String.to_integer(length_string), comparison_string}
+  end
+
+  defp as_text(value, _context) when is_binary(value), do: {:ok, value}
+
+  defp as_text(value, %ComparisonContext{} = context) do
+    %ComparisonContext{condition: condition, key: key} = context
+    attribute_name = UserCondition.comparison_attribute(condition)
+    condition_text = UserCondition.description(condition)
+
+    ConfigCatLogger.warning(
+      "Evaluation of condition (#{condition_text}) for setting '#{key}' may not produce the expected result " <>
+        "(the User.#{attribute_name} attribute is not a string value, thus it was automatically converted to " <>
+        "the string value '#{value}'). Please make sure that using a non-string value was intended.",
+      event_id: 3005
+    )
+
+    user_value_to_string(value)
+  end
+
+  defp user_value_to_string(nil), do: {:ok, nil}
+
+  defp user_value_to_string(%DateTime{} = dt) do
+    with {:ok, seconds} <- to_unix_seconds(dt) do
+      {:ok, to_string(seconds)}
+    end
+  end
+
+  defp user_value_to_string(%NaiveDateTime{} = naive) do
+    naive |> DateTime.from_naive!("Etc/UTC") |> user_value_to_string()
+  end
+
+  defp user_value_to_string(value) when is_list(value) do
+    with {:ok, list} <- to_string_list(value) do
+      {:ok, to_string(list)}
+    end
+  end
+
+  defp user_value_to_string(value), do: {:ok, to_string(value)}
+
+  defp to_float(value) when is_float(value), do: {:ok, value}
+  defp to_float(value) when is_integer(value), do: {:ok, value * 1.0}
+
+  defp to_float(value) when is_binary(value) do
+    value
+    |> String.replace(",", ".")
```

Comment:
As per the [specification](https://docs.google.com/document/d/1-Vi0YnX57P3gW1-_K_kaR1uZECQowajGJtbGwYvb3a8/#heading=h.lxt05eeqv4rp), leading/trailing whitespace in the user attribute value should be ignored in the case of number and date comparators.

According to my tests, `Float.parse()` doesn't allow leading/trailing whitespace, so we need to trim the string.

```suggestion
    |> String.trim()
    |> String.replace(",", ".")
```

Reply:
Fixed in 035cacb